### PR TITLE
Improved handling of micromega caches

### LINKED ISSUE #10772: Option to disable lia caching
No matter how [fast](https://github.com/coq/coq/pull/10765) or [slow](https://github.com/coq/coq/issues/10743#issuecomment-531215534) lia caching is, it would be good to have an option to turn off the lia cache. This would make it easier to understand benchmark results, enable us to test whether it actually improves performance at all, and simplify lia benchmarking.

/cc @fajb @thery @andres-erbsen

### PR DESCRIPTION
**Pull request by fajb**
The micromega caches are using the default Hashtbl.hash functions.
This PR is using a specialised hash function which traverses the whole key.
The hope is to reduce the amount of collisions that can be costly in certain extreme cases.
(The issue was identified as a side-effect of  #10743) 
The efficiency impact on common workloads needs to be assessed... 

The PR also adds options to disable caches.

**Kind:**  performance 

Fixes #10772

### CODE REVIEW

**Comment by ppedrot** [comment]
@fajb doesn't this need to be rebased now that the zify in ML PR has been merged?

**Comment by vbgl** [comment]
Benchmark result (old: dfff69ef604e02703575cb1cb15b2f77eda5f0f4; new: e3826ed50e3cb970d9c813c56ffbbf248292957c).

~~~
┌────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┐
│                        │      user time [s]      │                 CPU cycles                  │              CPU instructions               │
│                        │                         │                                             │                                             │
│           package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│ coq-mathcomp-ssreflect │   40.04   40.32 -0.69 % │      111296268021      111390372761 -0.08 % │      135193169178      135161117431 +0.02 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│         coq-coquelicot │   76.80   77.04 -0.31 % │      211818963486      211849300681 -0.01 % │      256861634563      256861967246 -0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│       coq-math-classes │  217.64  218.06 -0.19 % │      608158553041      608254427060 -0.02 % │      770475156003      770475372835 -0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│         coq-verdi-raft │ 1326.98 1328.64 -0.12 % │     3700138191483     3705456496001 -0.14 % │     5012804604747     5012942561077 -0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│        coq-lambda-rust │ 1539.19 1540.97 -0.12 % │     4288131811228     4294900466399 -0.16 % │     5686704140627     5684974256509 +0.03 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│           coq-compcert │  777.07  777.89 -0.11 % │     2166434349152     2166511379350 -0.00 % │     2895324352588     2895254025623 +0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│              coq-flocq │  247.20  247.43 -0.09 % │      685947747390      687902979050 -0.28 % │      858371348763      864639159702 -0.72 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│       coq-fiat-parsers │  650.49  651.07 -0.09 % │     1821949142560     1822481208801 -0.03 % │     2738445210873     2738393132498 +0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│              coq-color │  721.87  722.20 -0.05 % │     2018307898569     2017625911820 +0.03 % │     2403866754311     2403787770327 +0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│        coq-fiat-crypto │ 3825.32 3827.03 -0.04 % │    10614228898438    10621291004590 -0.07 % │    16637172780976    16655421224237 -0.11 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│          coq-fiat-core │  113.62  113.67 -0.04 % │      324005581998      323902362782 +0.03 % │      403112082487      403094993788 +0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│             coq-geocoq │ 1577.88 1578.29 -0.03 % │     4408084841067     4406057504377 +0.05 % │     6450751238969     6451148711376 -0.01 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│          coq-fourcolor │ 2165.07 2165.60 -0.02 % │     6035899825703     6036973254840 -0.02 % │    11360511135266    11360563573312 -0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│              coq-verdi │  120.49  120.51 -0.02 % │      334726873176      334695264191 +0.01 % │      420125211390      420149200952 -0.01 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│ coq-fiat-crypto-legacy │ 6256.75 6257.44 -0.01 % │    17410546462235    17408782954110 +0.01 % │    28270203573372    28261721326933 +0.03 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│            coq-unimath │ 4051.16 4050.96 +0.00 % │    11285436020881    11284944951249 +0.00 % │    20869071536445    20870078384799 -0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│ coq-mathcomp-odd-order │ 1439.43 1438.46 +0.07 % │     4013026877835     4009399527222 +0.09 % │     6858950554431     6858696318248 +0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│               coq-corn │ 1467.31 1465.32 +0.14 % │     4094118803404     4090588688215 +0.09 % │     6057147294134     6057122711446 +0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│            coq-bignums │   68.93   68.82 +0.16 % │      188812571017      188917298241 -0.06 % │      250878088394      250937270939 -0.02 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│   coq-mathcomp-algebra │  165.03  164.68 +0.21 % │      460005631899      459696801135 +0.07 % │      568657685968      568661117588 -0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│     coq-mathcomp-field │  230.50  230.00 +0.22 % │      642794865306      641331143932 +0.23 % │      917338516173      917278266346 +0.01 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│  coq-mathcomp-fingroup │   51.61   51.49 +0.23 % │      143760941521      143460572814 +0.21 % │      186969945407      186940419127 +0.02 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│ coq-mathcomp-character │  182.19  181.68 +0.28 % │      507552439880      506736385061 +0.16 % │      670584856456      670667519976 -0.01 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│  coq-mathcomp-solvable │  189.61  188.93 +0.36 % │      528136867705      526596444030 +0.29 % │      704824454565      704829766516 -0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│               coq-hott │  391.31  389.82 +0.38 % │     1063798760137     1062829016036 +0.09 % │     1611397338389     1611360373979 +0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│             coq-sf-plf │   44.60   44.26 +0.77 % │      125116227077      125134971684 -0.01 % │      161882987550      161847887156 +0.02 % │
└────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┘
~~~

**Comment by maximedenes** [comment]
So no improvement. @fajb do you have examples where such collisions occur?

**Comment by fajb** [comment]
@maximedenes , #10743 shows a big slow down.
The cache is 30 Mo big and takes many seconds (80s) to load with the default hash and less than a second with a full hash.
Maybe, we don't care such extreme cases...

**Comment by samuelgruetter** [comment]
> So no improvement. 

I'd rather say "no improvement in the cases for which lia works as of today". To decide whether to merge a new feature, we should not only do benchmarks on things which work as of today, but also benchmarks on things which are not being done as of today because it's just infeasably slow. A really performant lia could be a big enabler for things which were never done before.

**Comment by maximedenes** [comment]
> I'd rather say "no improvement in the cases for which lia works as of today". To decide whether to merge a new feature, we should not only do benchmarks on things which work as of today, but also benchmarks on things which are not being done as of today because it's just infeasably slow. A really performant lia could be a big enabler for things which were never done before.

Sure, I didn't mean that it was a reason not to merge the patch. I rather meant I was curious to see some problematic cases (I wasn't aware of the PR @fajb mentioned).

File: plugins/micromega/mutils.ml [hunk 1]
```diff
@@ -420,6 +427,52 @@ let command exe_path args vl =
...
+      fun helt ->
+      let rec hash acc = function
+        | Mc.Pc c -> helt (mix acc 1) c
+        | Mc.Pinj(p,c) -> hash (mix  (mix acc 1) (CoqToCaml.index p)) c
+        | Mc.PX(p1,i,p2) -> hash (hash (mix (mix acc 2) (CoqToCaml.index i)) p1) p2 in
+      hash
+
+    let hash_list : (int -> 'a -> int) -> int -> 'a list -> int =
+      fun helt ->
+      let rec hash acc = function
```
Comment (vbgl) [inline plugins/micromega/mutils.ml]: This looks like `fold_left`

File: plugins/micromega/mutils.ml [hunk 2]
```diff
@@ -420,6 +427,52 @@ let command exe_path args vl =
...
+(** Hashing utilities - borrowed from OCaml hash.c *)
+
+module Hash =
+  struct
+
+    module Mc = Micromega
+
+    let rotl32 x n = (x lsl n) lor (x lsr (32 - n))
+
+    let mix h d =
```
Comment (vbgl) [inline plugins/micromega/mutils.ml]: This could be `Hashset.Combine.combine`.

Comment (fajb) [reply plugins/micromega/mutils.ml]: Thanks. Done.

**Comment by fajb** [comment]
If the CI succeeds, I think the PR is ready...

File: doc/changelog/04-tactics/10765-micromega-caches.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+- Introduction of flags :flag:`Lia Cache`, :flag:`Nia Cache` and :flag:`Nra Cache`.
+  (`#10772 <https://github.com/coq/coq/issues/10772>`_ fixes, by Frédéric Besson).
```
Comment (Zimmi48) [inline doc/changelog/04-tactics/10765-micromega-caches.rst]: Is this what you meant?

```suggestion
  (`#10765 <https://github.com/coq/coq/pull/10765>`_,
  fixes `#10743 <https://github.com/coq/coq/issues/10743>`_
  and `#10772 <https://github.com/coq/coq/issues/10772>`_, by Frédéric Besson).
```

**Review by Zimmi48 — COMMENTED** [review]
I think the doc (and probably the changelog entry as well) should say clearly what gain is expected from these new flags (that are on by default, so there is no visible change by default).

File: plugins/micromega/coq_micromega.ml [hunk 4]
```diff
@@ -1978,13 +2015,38 @@ type provername = string * int option
...
+module MakeCache(T : sig type prover_option
+                         type coeff
+                         val  hash_coeff : int -> coeff -> int
+                 end) =
+  struct
+    module E =
+      struct
+        type t = T.prover_option * (T.coeff Mc.pol * Mc.op1) list
+
+        let equal = (=)
```
Comment (ppedrot) [inline plugins/micromega/coq_micromega.ml]: Can you write a dedicated equality function here? Parametric equality is known to introduce slowness, it's always better to write out your own equality cheks.

File: plugins/micromega/coq_micromega.ml [hunk 5]
```diff
@@ -1978,13 +2015,38 @@ type provername = string * int option
...
+
+        let hash  =
+          let hash_cstr  = Hash.(hash_pair (hash_pol T.hash_coeff) hash_any) in
+          Hash.( (hash_pair hash_any (List.fold_left hash_cstr)) 0)
```
Comment (ppedrot) [inline plugins/micromega/coq_micromega.ml]: Ideally the same here for hash.

**Review by ppedrot — APPROVED** [review]

File: plugins/micromega/coq_micromega.ml [hunk 6]
```diff
@@ -2018,18 +2018,22 @@ open Persistent_cache
...
+                         val  eq_coeff   : coeff -> coeff -> bool
+
+                     end) =
   struct
     module E =
       struct
         type t = T.prover_option * (T.coeff Mc.pol * Mc.op1) list
 
-        let equal = (=)
+        let equal = Hash.(eq_pair T.eq_prover_option (List.for_all2 (eq_pair (eq_pol T.eq_coeff) Hash.eq_op1)))
```
Comment (ppedrot) [inline plugins/micromega/coq_micromega.ml]: Use `CList.equal`, otherwise `for_all2` fails when the lengths differs.

**Comment by ppedrot** [comment]
@fajb I think you should squash the three commits into one, and fix the bootstrap of `micromega.ml` for the CI to go through. When this is done, I can assign and merge.

**Comment by Zimmi48** [comment]
> When this is done, I can assign and merge.

And the test-suite has been fixed, and my remarks on the doc have been answered :smile:

**Comment by ppedrot** [comment]
@Zimmi48 indeed, missed the doc comment. But the test-suite is precisely fixed by bootstrapping micromega!

**Comment by ppedrot** [comment]
Also, FTR, this is already critical for the stdlib. `theories/Numbers/Cyclic/Int63/Int63.v` is taking several minutes to compile on master and it's a PITA. I am surprised this was not observed beforehand.

**Comment by ppedrot** [comment]
@fajb what about the doc?

**Comment by fajb** [comment]
What about it?

**Comment by ppedrot** [comment]
@Zimmi48 asked for minor fixes, and they were not performed (https://github.com/coq/coq/pull/10765#pullrequestreview-292404425, https://github.com/coq/coq/pull/10765#discussion_r327603361).

**Comment by ppedrot** [comment]
@fajb ping

**Comment by fajb** [comment]
Sorry for the delay.
I do not thing amending the doc is the way forward.
I propose the following options:
- Keep it as it is
- Trash the Cache options (and not grant feature wish  #10772 )

I don't see myself what could be gained by disabling caches; so documenting makes no sense.

**Comment by ppedrot** [comment]
Personally I don't care about neither the option nor the documentation, but I'd like to get this PR *as fast as possible* into master. The ludicrous amount of time spent in a handful of files during stdlib compilation, which I perform several times a day, is driving me crazy.

**Comment by maximedenes** [comment]
Please don't introduce commands without documentation.

I don't see a problem with documenting a cache configuration option, TBH. In particular I do think it makes sense, with the use case described in #10772 for example (benchmarking).

**Comment by fajb** [comment]
They are documented. see micromega.rst

**Comment by maximedenes** [comment]
> They are documented. see micromega.rst

Ah yes, sorry. I meant that @Zimmi48 's comment could be addressed by mentioning the use case described in #10772 (benchmarking).

**Comment by ppedrot** [comment]
@maximedenes what's the conclusion then?

**Review by Zimmi48 — APPROVED** [review]
I don't get what was so hard in adding a sentence to the description of these flags, but I'm fine with the state of this PR as it is.

**Comment by thery** [comment]
@fajb I don't understand these options about cache. My main usage is to empty it.
```Unset Lia Cache.
   Set Lia Cache.```
is cleaning the cache or I still need to remove it by hand?

**Comment by fajb** [comment]
@Zimmi48 There is nothing hard but less is better.
@thery If the option is unset the cache is not used but still there. 
If you really want, you need to erase it, something you already can do it yourself : rm.
Ignoring requires some support.

To be frank, I have already exhausted my spare time on the issue. 
@thery , please, take the lead on this PR and implement the feature you wish for.

**Comment by thery** [comment]
@fajb ok no problem I will do the rm. Thanks!

**Pull request merged**